### ISSUE TITLE
unittest: execute tests in parallel

### ISSUE THREAD

**Issue by 68d74c6c-214f-4077-a72c-1bf43d7ec57c** [python/cpython#82054]
BPO | [37873](https://bugs.python.org/issue37873)
--- | :---
Nosy | @terryjreedy, @giampaolo, @ezio-melotti, @voidspace, @bharel, @tirkarthi, @donpellegrino

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-08-16.14:15:27.971>
labels = ['type-feature', 'library', '3.9']
title = 'unittest: execute tests in parallel'
updated_at = <Date 2020-09-02.09:57:36.866>
user = 'https://github.com/donpellegrino'
```

bugs.python.org fields:
```python
activity = <Date 2020-09-02.09:57:36.866>
actor = 'vstinner'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2019-08-16.14:15:27.971>
creator = 'user93448'
dependencies = []
files = []
hgrepos = []
issue_num = 37873
keywords = []
message_count = 5.0
messages = ['349864', '349938', '350080', '375126', '375128']
nosy_count = 7.0
nosy_names = ['terry.reedy', 'giampaolo.rodola', 'ezio.melotti', 'michael.foord', 'bar.harel', 'xtreak', 'user93448']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue37873'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-99637
* gh-103859
* gh-103927
* gh-103934

**Comment by gpshead** [comment]
An example of implementing sharding within a single unittest so they can run in parallel across whatever number of processes and machines you want is what we use at work: https://github.com/abseil/abseil-py/blob/v1.3.0/absl/testing/absltest.py#L2359 using the Bazel sharding protocol.

If `test.regrtest` supported using such a sharding protocol and unittest adopted that TestCase.getTestCaseNames implementation to compute the shards, we'd really pull in our long tail test time in CPython's own unittest suite.

> One of the ideas in the thread was to move `test.regrtest` parallel execution functionality into unittest. I think this would be good to have it in unittest like support in pytest for -j.

Yes. Though wider Python community wise you'll just find many people saying "just use pytest" in your project for -j support. I can't disagree with that.

I still want parallelization for CPython's own unittest suite (launched via `python -m test`) regardless of that. It'd help save core dev & contributor time as well as increasing buildbot and CI thruput.

**Comment by gpshead** [comment]
I'm leaving this open as there's plenty more opportunity to do things here (up to and including potentially making my full parallelism PR work reliably).  The just merged PR at least splits our longest test, test_asyncio, up into parallel sub-tests given it was already defined as a package of tests in separate files.

Refactoring `test_concurrent_futures.py` and the entire `_test_multiprocessing.py` suite and its trio of start-method runners into similar package of sub-test.py files would allow it to use the already committed pieces upon `make test` or `python -m test` when no list of tests is specified to further pull in the long tail on default run-everything CI, buildbot, and developer runs.

**Comment by gpshead** [comment]
With the refactorings @vstinner has done recently, we're close enough that I'm going to just close this issue.  See my now closed draft PR #99637 for the full parallelism implementation I started with and reasoning for why it likely isn't worth us chasing at this point.

Python users who want that parallelism in their own test suites can use `absltest` from https://pypi.org/project/absl-py/ or other frameworks that offer it.
